strawberryfields/compilers/gaussian_merge.py

```diff
@@ -0,0 +1,376 @@
+    Helper function that returns list of integers, which are the qumode indexes that are operated by op
+    """
+    return [reg.ind for reg in op.reg]
+
+
+def get_op_name(op):
+    """
+    Helper function that obtains the string name of an operation
+    """
+    return op.op.__class__.__name__
+
+
+class GaussianMerge(Fock):
+    """Compiler that merges adjacent Gaussian operations into a single symplectic transformation,
+    to reduce the depth of non-Gaussian programs.
+
+    As a result, the Gaussian operations that this compiler returns are :class:`~.ops.GaussianTransform`,
+    and :class:`~.ops.Displacement`. Meanwhile, non-Gaussian operations remain unchanged.
+
+    **Example:**
+
+    Consider the following Strawberry Fields program, compiled using the `'gaussian_merge'` compiler:
+
+    .. code-block:: python3
+
+        from strawberryfields.ops import Xgate, Zgate, BSgate, Rgate
+        import strawberryfields as sf
+
+        circuit = sf.Program(1)
+        with circuit.context as q:
+            Rgate(0.6)|q[0]
+            Rgate(0.2)|q[0]
+            Kgate(0.4)|q[0]
+            Rgate(0.1)|q[0]
+            Rgate(0.6)|q[0]
+            Dgate(0.01)|(q[0])
+
+        compiled_circuit = circuit.compile(compiler="gaussian_merge")
+
+    We can now print the compiled circuit, which has merged adjacent Gaussian
+    operations into singular :class:`~.GaussianTransform` operations:
+
+    >>> compiled_circuit.print()
+    GaussianTransform([[ 0.6967 -0.7174]
+      [0.7174  0.6967]]) | (q[0])
+    Kgate(0.4)|q[0]
+    GaussianTransform([[ 0.7648 -0.6442]
+      [0.6442  0.7648]]) | (q[0])
+    Dgate(0.01, 0) | (q[0])
+    """
+
+    short_name = "gaussian_merge"
+    primitives = {
+        # meta operations
+        "All",
+        "_New_modes",
+        "_Delete",
+        # state preparations
+        "Ket",
+        # measurements
+        "MeasureFock",
+        "MeasureHomodyne",
+        # single mode gates
+        "Dgate",
+        "Sgate",
+        "Rgate",
+        "Vgate",
+        "Kgate",
+        # two mode gates
+        "MZgate",
+        "sMZgate",
+        "BSgate",
+        "CKgate",
+        "S2gate",
+        "Interferometer",
+        "GaussianTransform",
+    }
+
+    decompositions = {
+        "GraphEmbed": {},
+        "BipartiteGraphEmbed": {},
+        "Gaussian": {},
+        "Pgate": {},
+        "CXgate": {},
+        "CZgate": {},
+        "Xgate": {},
+        "Zgate": {},
+        "Fouriergate": {},
+    }
+
+    prep_states = ["Ket"]
+
+    gaussian_ops = [
+        "Dgate",
+        "BSgate",
+        "S2gate",
+        "Sgate",
+        "GaussianTransform",
+        "Rgate",
+        "Interferometer",
+        "MZgate",
+        "sMZgate",
+    ]
+
+    def __init__(self):
+        self.curr_seq = None
+        self.DAG = None
+        self.new_DAG = None
+
+    def compile(self, seq, registers):
+        """Attempt to merge Gaussian operations into Gaussian Transforms in a hybrid program.
```

Comment:
It might be good here @federico0112 to outline the algorithm, that is, have a bullet point list of steps that the compilation performs, as well as the names of the methods that perform that part of the logic.

This will help future developers piece together the logical flow :)

Reply:
I addded it to the docstring of merge_a_gaussian_op cause that is where most of the logic flow is done. Let me know if it doesnt make sense or its wrongly formatted.